### ISSUE TITLE
2.5D view

### ISSUE THREAD

**Issue by Goldsong** [aroberge/reeborg#120]
Since the issue of graphics has been on the table lately, maybe we should consider how this will work before we embark in the process of drawing tiles, i.e., to consider the app after the graphics are added.

The issue is the way that the user perceives the grid.  I discussed this with André a while back; we were used to different representations of the grid. It is possible to see the grid as representing a vertical surface, with the bottom of the grid being the floor. This representation allows missions such as the robot climbing a mountain or a staircase.  The original images of reeborg, of a humanoid robot, follow this convention. In this case the grid directions are perceived as left, right, up and down.

A second way to see the grid is as a representation of the floor, i.e., as the world seen from a birds-eye point of view. This representation allows missions such as solving a maze. In this case, the directions of the grid are east, west, north and south. The images of the red 4-wheeled robot seen from above follow this convention.

Although these are two different ways to see the grid, the students have little problem jumping between missions that use them; one mission might be a solving a maze followed by a mission that asks the robot to climb a mountain. However, from the point of view of graphics, we would need two different images: say, a tile of a tree seen from the side, drawn to support the side-view representation, would be different from a tile of a tree seen from above, drawn to support the birds-eye view representation. Why not mix them?

In the the studio code pointer example, the representation is an orthographic 2.5D projection, not a side view but not a birds-eye view either.  Using this representation would eliminate the need of having 2 types of tiles. The only downside that I see is that drawing anything in an orthographic projection is more difficult than drawing it from the side or from above, e.g., we'd need to draw 3-D looking cubes instead of squares.  An interesting byproduct of this, though, is that the 2.5D tiles serve as a mediator between side-view and birds-eye view objects. For example, in the example at studio code, the flower is drawn from the birds-eye view, from above, but the bee is drawn from the side view; however, neither of them look out of place because the tiles of the trees drawn in 2.5D 'average' their views.

In summary... this change is purely cosmetic; no code changes needed. Should we consider moving towards a 2.5D world, drawing 2.5D tiles?

**Comment by aroberge** [comment]
I really like the idea of 2.5D tiles.

My idea on this topic is to build a collection of images (objects, robots and tiles) which could be used to create worlds.  Whether 2.5D images are used or not should be left to the world creator, and the program should be flexible enough to accommodate this.    

Right now, only walls (between tiles) can block a path.  A while ago, I added a "mud" tile which was undetectable by Reeborg ... but would cause a fatal error if Reeborg were to step on it.    What I am thinking is to have tiles that can either be just for background effects, or they could present obstacles to Reeborg - some of which that he perhaps could not detect with its senses (like the current mud tiles) until it is too late.   Some could perhaps be drawn on the foreground (e.g. Reeborg entering a tunnel).

Whether all of this is done with 2.5D tiles (which is probably the best visual option) or better looking 2D tiles and objects is going to be determined by the success in finding the right images.

**Comment by Goldsong** [comment]
> On May 19, 2015, at 12:22 PM, André Roberge <notifications@github.com> wrote:
> 
> I really like the idea of 2.5D tiles.
> 
> 
=-)

nice. otherwise I’d wasted a couple of hours preparing a sample of what the square robot might look in 2.5D.  Here is a sample:


> My idea on this topic is to build a collection of images (objects, robots and tiles) which could be used to create worlds. Whether 2.5D images are used or not should be left to the world creator, and the program should be flexible enough to accommodate this.
> 
yes
> Right now, only walls (between tiles) can block a path. A while ago, I added a "mud" tile which was undetectable by Reeborg ... but would cause a fatal error if Reeborg were to step on it.
> 
I used them a few times, particularly in mazes.
> What I am thinking is to have tiles that can either be just for background effects, or they could present obstacles to Reeborg - some of which that he perhaps could not detect with its senses (like the current mud tiles) until it is too late. Some could perhaps be drawn on the foreground (e.g. Reeborg entering a tunnel).
> 

very nice.. a possible tile could be a door that open and closes (with it respective sound).  Many video games of the 70s and 80s were multi-level puzzles and the way that the character moved from one level to the next one was getting to the door, e.g., Super Mario Bros.
> Whether all of this is done with 2.5D tiles (which is probably the best visual option) or better looking 2D tiles and objects is going to be determined by the success in finding the right images.
> 
yes.. the best quality images will get used the most

**Comment by aroberge** [comment]
@Goldsong The sample image does not show; perhaps images can only be shown when replying to the issue on github.

**Comment by aroberge** [comment]
On Tue, May 19, 2015 at 4:58 PM, Andres Castano <notifications@github.com>
wrote:

>
>
>
> very nice.. a possible tile could be a door that open and closes (with it
> respective sound). Many video games of the 70s and 80s were multi-level
> puzzles and the way that the character moved from one level to the next one
> was getting to the door, e.g., Super Mario Bros.
>
> Hmmm... I could see how one could write some post-code that checks if the
goal is satisfied and, if so, updates automatically the permalink to load a
new world to solve. Interesting idea...

**Comment by Goldsong** [comment]
hum.. I tried to upload the image in 3 different ways, none of which worked.

**Comment by aroberge** [comment]
Send by email and I'll see what I can do?

**Comment by aroberge** [comment]
![25d-orthographic-view](https://cloud.githubusercontent.com/assets/629698/7713337/8a8bcc18-fe4d-11e4-86ec-77c24633712a.png)

**Comment by aroberge** [comment]
This can be done at any time with someone providing suitable tiles.  Not critical enough to keep open as an active issue.
